tests/Microsoft.Experimental.Collections.Tests/DictionarySlimTests.cs

```diff
@@ -0,0 +1,387 @@
+            var rd = new DictionarySlim<ulong, int>();
+            var d = new Dictionary<ulong, int>();
+            var size = 1000;
+
+            for (int i = 0; i < size; i++)
+            {
+                var k = (ulong)rand.Next(100) + 23;
+                var v = rand.Next();
+
+                rd[k] += v;
+
+                if (d.TryGetValue(k, out int t))
+                    d[k] = t + v;
+                else
+                    d[k] = v;
+            }
+
+            Assert.Equal(d.Count, rd.Count);
+            Assert.Equal(d.OrderBy(i => i.Key), (rd.OrderBy(i => i.Key)));
+            Assert.Equal(d.OrderBy(i => i.Value), (rd.OrderBy(i => i.Value)));
+        }
+
+        [Fact]
+        public void DictionarySlimVersusDictionary_AllCollisions()
+        {
+            var rand = new Random(333);
+            var rd = new DictionarySlim<Collider, int>();
+            var d = new Dictionary<Collider, int>();
+            var size = rand.Next(1234);
+
+            for (int i = 0; i < size; i++)
+            {
+                if (rand.Next(5) != 0)
+                {
+                    var k = C(rand.Next(100) + 23);
+                    var v = rand.Next();
+
+                    rd[k] += v;
+
+                    if (d.TryGetValue(k, out int t))
+                        d[k] = t + v;
+                    else
+                        d[k] = v;
+                }
+
+                if (rand.Next(3) == 0 && d.Count > 0)
+                {
+                    var el = GetRandomElement(d);
+                    Assert.True(rd.Remove(el));
+                    Assert.True(d.Remove(el));
+                }
+            }
+
+            Assert.Equal(d.Count, rd.Count);
+            Assert.Equal(d.OrderBy(i => i.Key), (rd.OrderBy(i => i.Key)));
+            Assert.Equal(d.OrderBy(i => i.Value), (rd.OrderBy(i => i.Value)));
+        }  
+
+        private TKey GetRandomElement<TKey, TValue>(IDictionary<TKey, TValue> d)
+        {
+            int index = 0;
+            var rand = new Random(42);
+            foreach(var entry in d)
+            {
+                if (rand.Next(d.Count) == 0 || index == d.Count - 1)
+                {
+                    return entry.Key;
+                }
+
+                index++;
+            }
+
+            throw new InvalidOperationException(); 
+        }             
+
+        [DebuggerStepThrough]
+        internal Collider C(int val) => new Collider(val);
+
+        [Fact]
+        public void Collision()
+        {
+            var d = new DictionarySlim<Collider, int>();
+            d[C(5)] = 3;
+            d[C(7)] = 9;
+            d[C(10)] = 11;
+            Assert.Equal(3, d.GetValueOrDefault(C(5)));
+            Assert.Equal(9, d.GetValueOrDefault(C(7)));
+            Assert.Equal(11, d.GetValueOrDefault(C(10)));
+            d[C(23)]++;
+            d[C(23)] += 3;
+            Assert.Equal(4, d[C(23)]);
+        }
+
+        internal class UsedKey : IEquatable<UsedKey>
```

Comment:
Is there a better name that suggests what this is for? Eg., `KeyTrackingIEquatableEquals`

Reply:
Or `InstrumentedKey` ?

Reply:
Changed